salt/modules/boto3_asg.py

```diff
@@ -0,0 +1,946 @@
+
+.. versionadded:: 2014.7.0
+
+:configuration: This module accepts explicit autoscale credentials but can also
+    utilize IAM roles assigned to the instance through Instance Profiles.
+    Dynamic credentials are then automatically obtained from AWS API and no
+    further configuration is necessary. More Information available at:
+
+    .. code-block:: text
+
+        http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html
+
+    If IAM roles are not used you need to specify them either in a pillar or
+    in the minion's config file:
+
+    .. code-block:: yaml
+
+        asg.keyid: GKTADJGHEIQSXMKKRBJ08H
+        asg.key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+
+    A region may also be specified in the configuration:
+
+    .. code-block:: yaml
+
+        asg.region: us-east-1
+
+    If a region is not specified, the default is us-east-1.
+
+    It's also possible to specify key, keyid and region via a profile, either
+    as a passed in dict, or as a string to pull from pillars or minion config:
+
+    .. code-block:: yaml
+
+        myprofile:
+            keyid: GKTADJGHEIQSXMKKRBJ08H
+            key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+:depends: boto
+:depends: boto3
+'''
+# keep lint from choking on _get_conn and _cache_id
+#pylint: disable=E0602
+
+# Import Python libs
+from __future__ import absolute_import, print_function, unicode_literals
+import datetime
+import logging
+import sys
+import time
+import email.mime.multipart
+
+log = logging.getLogger(__name__)
+DATE_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+# Import third party libs
+from salt.ext import six
+try:
+    import boto
+    import boto.ec2
+    import boto.ec2.instance
+    import boto.ec2.blockdevicemapping as blockdevicemapping
+    import boto3
+    import boto.ec2.autoscale as autoscale
+    logging.getLogger('boto').setLevel(logging.CRITICAL)
+    import boto3  # pylint: disable=unused-import
+    from botocore.exceptions import ClientError
+    logging.getLogger('boto3').setLevel(logging.CRITICAL)
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+
+
+# Import Salt libs
+import salt.utils.compat
+import salt.utils.json
+import salt.utils.odict as odict
+import salt.utils.versions
+
+
+def __virtual__():
+    '''
+    Only load if boto libraries exist.
+    '''
+    has_boto_reqs = salt.utils.versions.check_boto_reqs()
+    if has_boto_reqs is True:
+        __utils__['boto.assign_funcs'](__name__, 'asg', module='ec2.autoscale', pack=__salt__)
+        setattr(sys.modules[__name__], '_get_ec2_conn',
+                __utils__['boto.get_connection_func']('ec2'))
+    return has_boto_reqs
+
+
+def __init__(opts):
+    salt.utils.compat.pack_dunder(__name__)
+    if HAS_BOTO:
+        __utils__['boto3.assign_funcs'](
+            __name__, 'autoscaling',
+            get_conn_funcname='_get_conn_autoscaling_boto3')
+
+
+def exists(name, region=None, key=None, keyid=None, profile=None):
```

Comment:
should we be including these methods that haven't been updated here given they will almost certainly be broken still?  (ie using `boto.get_all_groups`)

Reply:
Yep agreed that they should be removed, but I was planning on verifying that we've patched all that needs to be patched before removing them.

Reply:
These have now been removed now that the initial modifications have been tested.